File: Ruby/textUI.rb

require 'io/console'
require_relative 'Directions'

module UI

  class TextUI

    #https://gist.github.com/acook/4190379
    def read_char
      STDIN.echo = false
      STDIN.raw!
    
      input = STDIN.getc.chr
      if input == "\e" 
        input << STDIN.read_nonblock(3) rescue nil
        input << STDIN.read_nonblock(2) rescue nil
      end
    ensure
      STDIN.echo = true
      STDIN.cooked!
    
      return input
    end

    def next_move
      print "Where? "
      got_input = false
      while (!got_input)
        c = get.chomp
        case c
          when "\e[W"
            puts "UP ARROW"
            output = Irrgarten::Directions::UP
            got_input = true
          when "\e[S"
            puts "DOWN ARROW"
            output = Irrgarten::Directions::DOWN
            got_input = true
          when "\e[D"
            puts "RIGHT ARROW"
            output = Irrgarten::Directions::RIGHT
            got_input = true
          when "\e[A"
            puts "LEFT ARROW"
            output = Irrgarten::Directions::LEFT
            got_input = true
          when "\u0003"
            puts "CONTROL-C"
            got_input = true
            exit(1)
          else
            #Error
        end
      end
      output
    end

    def show_game(game_state)

      puts game_state.get_labyrinth
      puts game_state.get_players
      puts game_state.get_monsters
      puts game_state.get_currentPlayer
      puts game_state.get_winner
      puts game_state.get_log

    end

  end # class   

end # module   

File: Irrgarten-Java/src/UI/TextUI.java
/**
 * @file UI/TextUI.java
 * @author Álvaro Maldonado Medina, grupo: 2D-D3
 */

package UI;

import irrgarten.Directions;
import irrgarten.GameState;
import java.util.Scanner;

/**
 * Interfaz de usuario de texto para el juego 
 */

public class TextUI {
    
    private static Scanner in = new Scanner(System.in);
    
    /**
     * Lee un carácter desde la entrada estándar
     * @return El carácter que ingresa el usuario por teclado
     */
    
    private char readChar() {
        String s = in.nextLine();     
        return s.charAt(0);
    }
    
    /**
     * Solicita al usuario el próximo movimiento y devuelve la dirección seleccionada
     * @return La dirección seleccionada por el usuario
     */
    
    public Directions nextMove() {
        System.out.print("Where? ");
        
        Directions direction = Directions.DOWN;
        boolean gotInput = false;
        
        while (!gotInput) {
            char c = readChar();
            switch(c) {
                case 'w':
                    System.out.print(" UP\n\n");
                    direction = Directions.UP;
                    gotInput = true;
                    break;
                case 's':
                    System.out.print(" DOWN\n\n");
                    direction = Directions.DOWN;
                    gotInput = true;
                    break;
                case 'd':
                    System.out.print("RIGHT\n\n");
                    direction = Directions.RIGHT;
                    gotInput = true;
                    break;
                case 'a':
                    System.out.print(" LEFT\n\n");
                    direction = Directions.LEFT;
                    gotInput = true;    
                    break;
            }
        }   
        return direction;
    }
    
    /**
     * Muestra el estado actual del juego en la terminal
     * @param gameState instancia que llama y muestra el GameState en la terminal
     */
    
    public void showGame(GameState gameState) {  
        
        System.out.println(gameState.getLabyrinth());
        System.out.println("Lista de jugadores: "+gameState.getPlayers());
        System.out.println("Lista de monstruos: "+gameState.getMonsters());
        System.out.println("Jugador actual: "+gameState.getCurrentPlayer());
        System.out.println("Ganador?: "+gameState.getWinner());
        System.out.println(gameState.getLog());
  
    }
    
}